--- python/CMakeLists.txt	2016-05-24 22:05:50.698570824 -0700
+++ ../new/python/CMakeLists.txt	2016-05-24 22:05:57.774570529 -0700
@@ -12,7 +12,7 @@
 
 add_custom_target(pytarget ALL DEPENDS ${OUTPUT}/timestamp)
 
-install(CODE "execute_process(COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} install --record files.txt)")
+#install(CODE "execute_process(COMMAND ${PYTHON_EXECUTABLE} ${SETUP_PY} install --record files.txt)")
 
 if (ENABLE_TESTING)
     #add_test (pytest ${PYTHON_EXECUTABLE} "${CMAKE_CURRENT_SOURCE_DIR}/test_pycryptosat.py")
