diff --git a/configuration/cxx14.sh b/configuration/cxx14.sh
index e974ab6..65cd94e 100755
--- a/configuration/cxx14.sh
+++ b/configuration/cxx14.sh
@@ -51,7 +51,7 @@ int main() {
 EOF
 
 # First try with no compiler flag...
-"$CXX"  language-version.C  -o language-version  >> LogFile  2>& 1 
+"$CXX" $CXXFLAGS language-version.C  -o language-version  >> LogFile  2>& 1 
 if [ $? -ne 0 ]
 then
     echo "ERROR: compilation unexpectedly failed; is $CXX a c++ compiler?   $SCRIPT_NAME" > /dev/stderr
@@ -76,7 +76,7 @@ fi
 # Compilation without flag is not C++14 standard; try with -std=c++14
 
 CXX14="-std=c++14"
-"$CXX"  $CXX14  language-version.C  -o language-version  >> LogFile  2>& 1 
+"$CXX" $CXXFLAGS  $CXX14  language-version.C  -o language-version  >> LogFile  2>& 1 
 if [ $? -ne 0 ]
 then
     echo "ERROR: compilation with flag $CXX14 failed   $SCRIPT_NAME"                      > /dev/stderr
diff --git a/configuration/gmp-version.sh b/configuration/gmp-version.sh
index 4fe33b8..2a17749 100755
--- a/configuration/gmp-version.sh
+++ b/configuration/gmp-version.sh
@@ -74,7 +74,7 @@ EOF
 
 # Use c++ compiler specified in CXX; no need to specify libgmp as all info is in header file!!
 echo "$CXX -I"$EXTLIB_DIR_FULL/include"  test-gmp-version.C  -o test-gmp-version"  > LogFile
-$CXX -I"$EXTLIB_DIR_FULL/include"  test-gmp-version.C  -o test-gmp-version  >> LogFile 2>&1
+$CXX $CXXFLAGS -I"$EXTLIB_DIR_FULL/include"  test-gmp-version.C  -o test-gmp-version  >> LogFile 2>&1
 
 # Check whether compilation failed; if so, complain.
 if [ $? -ne 0 ]
diff --git a/configuration/shell-fns.sh b/configuration/shell-fns.sh
index 46d8155..3687819 100755
--- a/configuration/shell-fns.sh
+++ b/configuration/shell-fns.sh
@@ -27,13 +27,13 @@ mktempdir()
 echounderline()
 {
   echo "$*"
-  echo "$*" | tr "\040-\377" "[-*]"
+  echo "$*" | tr "\040-\377" "-"
 }
 
 echobox()
 {
   mesg=">>>>  $*  <<<<"
-  dashes=`echo "$mesg" | tr "\040-\377" "[-*]"`
+  dashes=`echo "$mesg" | tr "\040-\377" "-"`
   echo "$dashes"
   echo "$mesg"
   echo "$dashes"
@@ -42,7 +42,7 @@ echobox()
 echoerror()
 {
   mesg=">>>>>  $*  <<<<<"
-  equals=`echo "$mesg" | tr "\040-\377" "[=*]"`
+  equals=`echo "$mesg" | tr "\040-\377" "="`
   echo "$equals"
   echo "$mesg"
   echo "$equals"
diff --git a/configuration/verify-compiler.sh b/configuration/verify-compiler.sh
index b5f5a95..8c28263 100755
--- a/configuration/verify-compiler.sh
+++ b/configuration/verify-compiler.sh
@@ -58,16 +58,6 @@ int main()
 }
 EOF
 
-# Try plain compiler (without CXXFLAGS):
-$CXX test-compiler-type.C -o test-compiler-type  > LogFile  2>&1
-if [ $? -ne 0 -o \! -f test-compiler-type -o \! -x test-compiler-type ]
-then
-  echo "ERROR: Are you sure \"$CXX\" is a C++ compiler?   $SCRIPT_NAME"  > /dev/stderr
-  echo "LOGFILE: $TMP_DIR/LogFile   $SCRIPT_NAME"                        > /dev/stderr
-  exit 1
-fi
-/bin/rm test-compiler-type  # not necessary, just being tidy :-)
-
 # Try compiler with CXXFLAGS:
 $CXX $CXXFLAGS test-compiler-type.C -o test-compiler-type  > LogFile  2>&1
 if [ $? -ne 0 -o \! -f test-compiler-type -o \! -x test-compiler-type ]
diff --git a/include/CoCoA/MakeUnifiedHeader.sh b/include/CoCoA/MakeUnifiedHeader.sh
index e0e713b..156461b 100755
--- a/include/CoCoA/MakeUnifiedHeader.sh
+++ b/include/CoCoA/MakeUnifiedHeader.sh
@@ -33,7 +33,7 @@ fi
 echobox()
 {
   mesg=">>>>  $*  <<<<"
-  dashes=`echo "$mesg" | tr "\040-\377" "[-*]"`
+  dashes=`echo "$mesg" | tr "\040-\377" "-"`
   echo "$dashes"
   echo "$mesg"
   echo "$dashes"
diff --git a/include/CoCoA/TmpGPoly.H b/include/CoCoA/TmpGPoly.H
index 4c4d51e..b57fc17 100644
--- a/include/CoCoA/TmpGPoly.H
+++ b/include/CoCoA/TmpGPoly.H
@@ -29,6 +29,7 @@
 #include "CoCoA/TmpGTypes.H"
 #include "CoCoA/utils.H"
 
+#include<functional>
 #include<list>
 // using std::list; // for GPolyList; GPolyPtrList;
 #include<vector>
@@ -46,6 +47,12 @@ enum ClearMarker {clear};
 
 class GPoly;
 
+extern bool handlersEnabled;
+extern std::function<void(ConstRefRingElem p, ConstRefRingElem q, ConstRefRingElem s)> sPolyHandler;
+extern std::function<void(ConstRefRingElem p)> reductionStartHandler;
+extern std::function<void(ConstRefRingElem q)> reductionStepHandler;
+extern std::function<void(ConstRefRingElem r)> reductionEndHandler;
+
 
 typedef std::list<GPoly> GPolyList;
 
diff --git a/src/AlgebraicCore/CheckSourceFiles.sh b/src/AlgebraicCore/CheckSourceFiles.sh
index b669e0f..cb0a66c 100755
--- a/src/AlgebraicCore/CheckSourceFiles.sh
+++ b/src/AlgebraicCore/CheckSourceFiles.sh
@@ -24,7 +24,7 @@ do
 done
 sort "$TMPFILE2" > "$TMPFILE1"
 
-/bin/ls -d ./*.C | cut -b 3- > "$TMPFILE2"
+/bin/ls -d ./*.C | cut -b 3- | sort > "$TMPFILE2"
 
 cmp "$TMPFILE1" "$TMPFILE2" > /dev/null
 if [ $? = 0 ]
diff --git a/src/AlgebraicCore/PolyRing-content.C b/src/AlgebraicCore/PolyRing-content.C
index c5d0a0a..36d6d2b 100644
--- a/src/AlgebraicCore/PolyRing-content.C
+++ b/src/AlgebraicCore/PolyRing-content.C
@@ -33,6 +33,7 @@
 #include "CoCoA/convert.H"
 #include "CoCoA/error.H"
 #include "CoCoA/utils.H"  // for len
+#include "CoCoA/TmpGPoly.H"
 
 
 #include <vector>
@@ -128,7 +129,9 @@ namespace CoCoA
   {
     const PolyRing Rx = owner(f);
     RingElem ans(Rx);
+    if (handlersEnabled) reductionStartHandler(f);
     Rx->myMonic(raw(ans), raw(f));
+    if (handlersEnabled) reductionEndHandler(ans);
     return ans;
   }
 
diff --git a/src/AlgebraicCore/SparsePolyOps-reduce.C b/src/AlgebraicCore/SparsePolyOps-reduce.C
index 30e407a..b8789e2 100644
--- a/src/AlgebraicCore/SparsePolyOps-reduce.C
+++ b/src/AlgebraicCore/SparsePolyOps-reduce.C
@@ -145,6 +145,7 @@ degree HereForProfilingOnlyWDeg(ConstRefPPMonoidElem cofactor1)
       s->myUpdate(F, *g);
       //      std::cout << "ReduceActiveLM after s->myUpdate" << std::endl;
       F->myReduce(poly(*g), NumTerms(*g));
+      if ( handlersEnabled ) reductionStepHandler(poly(*g));
     }//while
   }//ReduceActiveLM
 
@@ -283,6 +284,7 @@ degree HereForProfilingOnlyWDeg(ConstRefPPMonoidElem cofactor1)
   void GPoly::myReduce(const Reductors& theReductors)
   {
     if ( IsZero(*this) ) return;
+    if ( handlersEnabled ) reductionStartHandler(myPoly());
     ReductionCog F = ChooseReductionCogGeobucket(myGRingInfo());
     //    std::cout << "myreduce1: F is " << F << std::endl;
     F->myAssignReset(myPolyValue, myNumTerms); // myPolyValue gets 0
@@ -298,6 +300,7 @@ degree HereForProfilingOnlyWDeg(ConstRefPPMonoidElem cofactor1)
     if ( !IsZero(*this) && !IsOne(myLCValue) ) // makes myPolyValue monic
       if ( myGRingInfo().myCoeffRingType()==CoeffEncoding::Field )
         myGRingInfo().myNewSPR()->myDivByCoeff(raw(myPolyValue), raw(myLCValue));
+    if ( handlersEnabled ) reductionEndHandler(myPoly());
     // if CoeffEncoding::Field myRelease does NOT make poly monic
     // if CoeffEncoding::FrFldOfGCDDomain myRelease makes poly content free
   }
diff --git a/src/AlgebraicCore/TmpGPoly.C b/src/AlgebraicCore/TmpGPoly.C
index 3163ca9..c10b8c2 100644
--- a/src/AlgebraicCore/TmpGPoly.C
+++ b/src/AlgebraicCore/TmpGPoly.C
@@ -47,6 +47,11 @@ using std::vector;
 namespace CoCoA
 {  
 
+  bool handlersEnabled = false;
+  std::function<void(ConstRefRingElem p, ConstRefRingElem q, ConstRefRingElem s)> sPolyHandler = nullptr;
+  std::function<void(ConstRefRingElem p)> reductionStartHandler = nullptr;
+  std::function<void(ConstRefRingElem q)> reductionStepHandler = nullptr;
+  std::function<void(ConstRefRingElem r)> reductionEndHandler = nullptr;
 
   //---------class GPoly-------------------------------
 
@@ -291,7 +296,10 @@ void GPoly::myUpdateLenLPPLCDegComp()
     if (the_gp.IsInputPoly())
       myPolyValue = poly(the_gp.myFirstGPoly());
     else
+    {
       myPolySetSPoly(the_gp.myFirstGPoly(), the_gp.mySecondGPoly());
+      if ( handlersEnabled ) sPolyHandler(the_gp.myFirstGPoly().myPoly(), the_gp.mySecondGPoly().myPoly(), myPoly());
+    }
     myUpdateLenLPPLCDegComp();
     myAge = the_age;
     // MAX: do these things only if necessary.
